torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/math_ops.py

```diff
@@ -0,0 +1,69 @@
+import torch
+from torch import Tensor
+from torch.distributed._shard.sharded_tensor import (
+    ShardedTensor,
+)
+from torch.distributed._shard.sharding_spec import ChunkShardingSpec
+from torch.distributed._shard.sharding_spec.api import custom_sharding_spec_op
+from torch.distributed._shard.sharded_tensor._ops.math_ops import binary_math_op_impl
+
+def register_math_op(op):
+    @custom_sharding_spec_op(ChunkShardingSpec, op)
+    def binary_math_op(types, args=(), kwargs=None):
+        """
+        Handles ``__torch_function__`` dispatch for the binary math ops
+        such as `torch.add`, `torch.mul`, `torch.div`, etc.
+        This method computes on ShardedTensor
+        """
+        if len(args) != 2:
+            raise ValueError("Only support binary math op on ShardedTensor for now!")
+        lhs = args[0]
+        rhs = args[1]
+        pg = lhs._process_group if isinstance(lhs, ShardedTensor) else rhs._process_group
+        # Validate types
+        if isinstance(lhs, ShardedTensor) and isinstance(rhs, ShardedTensor):
+            lhs_spec = lhs.sharding_spec()
+            rhs_spec = rhs.sharding_spec()
+            if not isinstance(lhs_spec, ChunkShardingSpec) or not isinstance(rhs_spec, ChunkShardingSpec):
+                raise TypeError("Only ShardedTensor with ChunkShardingSpec supports"
+                                " two ShardedTensor together")
+
+            if lhs.size() == rhs.size() and lhs_spec.dim == rhs_spec.dim:
+                # perform local element-wise math op
+                res = op(lhs.local_tensor(), rhs.local_tensor())
+                return ShardedTensor._init_from_local_tensor(
+                    res,
+                    lhs_spec,
+                    lhs.size(),  # type: ignore[arg-type]
+                    process_group=pg)
+            else:
+                raise RuntimeError("Implicit broadcasting not supported yet!")
```

Comment:
Nit: we also don't support when sharding dim is different, right?

Reply:
Broadcasting probably covers the sharding dim mismatch case too since in that case you probably need to do broadcasting across ranks.